llvm/include/llvm/CAS/Utils.h

```diff
@@ -0,0 +1,33 @@
+//===- llvm/CAS/Utils.h -----------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CAS_UTILS_H
+#define LLVM_CAS_UTILS_H
+
+#include "llvm/Support/Error.h"
+
+namespace llvm {
+namespace cas {
+
+class CASDB;
+class CASID;
+class NamedTreeEntry;
+
+/// Visit each file entry in order, returning an error from \p Callback to stop
+/// early.
+///
+/// The \p NamedTreeEntry, that the \p Callback receives, points to a name
+/// string that may not live beyond the return of the callback function.
+Error walkFileTreeRecursively(
+    CASDB &CAS, CASID ID,
+    function_ref<Error(const NamedTreeEntry &, bool IsEmpty)> Callback);
```

Comment:
I wonder if this should be an `Optional<bool> IsTreeEmpty`, only set when it's valid. Otherwise it looks like a flag for an empty file.

Maybe this should really be another type (`DetailedTreeEntry`?) that wraps `NamedTreeEntry` and includes size info... could change it later if that turns out to be useful.

Reply:
On second thought the `IsEmpty` parameter is very specific to accommodating `listTreeRecursively()`, which wants to print an empty tree, but a caller could check whether a tree is empty in the callback anyway using `CAS.getTree()`.
Reporting that a tree is empty in the callback doesn't seem widely useful, I think I'll remove this parameter and have `listTreeRecursively()` check for it.